File: frontend/src/types/template.ts
/**
 * Template types aligned with backend schemas
 */

// Template placeholder
export interface TemplatePlaceholder {
    key: string
    label: string
    type: 'text' | 'number' | 'date' | 'select' | 'textarea'
    required?: boolean
    default?: string
    options?: string[]
    description?: string
}

// Template interface
export interface Template {
    id: string
    name: string
    description?: string
    content: string
    category?: string
    placeholders: TemplatePlaceholder[]
    is_active: boolean
    custom_metadata: Record<string, any>
    created_by: string
    created_at: string
    updated_at?: string
}

// Template list item
export interface TemplateListItem {
    id: string
    name: string
    description?: string
    category?: string
    is_active: boolean
    created_at: string
}

// Template creation
export interface TemplateCreate {
    name: string
    description?: string
    content: string
    category?: string
    placeholders?: TemplatePlaceholder[]
    metadata?: Record<string, any>
}

// Template update
export interface TemplateUpdate {
    name?: string
    description?: string
    content?: string
    category?: string
    placeholders?: TemplatePlaceholder[]
    is_active?: boolean
    metadata?: Record<string, any>
}

// Generate from template request
export interface GenerateFromTemplateRequest {
    template_id: string
    title: string
    values: Record<string, string>
    description?: string
    metadata?: Record<string, any>
}

// Template categories
export const TEMPLATE_CATEGORIES = [
    { value: 'nda', label: 'Non-Disclosure Agreement' },
    { value: 'service_agreement', label: 'Service Agreement' },
    { value: 'employment', label: 'Employment Contract' },
    { value: 'purchase_order', label: 'Purchase Order' },
    { value: 'lease', label: 'Lease Agreement' },
    { value: 'partnership', label: 'Partnership Agreement' },
    { value: 'vendor', label: 'Vendor Agreement' },
    { value: 'other', label: 'Other' }
] as const

export type TemplateCategory = typeof TEMPLATE_CATEGORIES[number]['value']

// Category display helpers
export const CATEGORY_COLORS: Record<string, string> = {
    nda: 'bg-purple-100 text-purple-800 dark:bg-purple-900 dark:text-purple-200',
    service_agreement: 'bg-blue-100 text-blue-800 dark:bg-blue-900 dark:text-blue-200',
    employment: 'bg-green-100 text-green-800 dark:bg-green-900 dark:text-green-200',
    purchase_order: 'bg-yellow-100 text-yellow-800 dark:bg-yellow-900 dark:text-yellow-200',
    lease: 'bg-orange-100 text-orange-800 dark:bg-orange-900 dark:text-orange-200',
    partnership: 'bg-pink-100 text-pink-800 dark:bg-pink-900 dark:text-pink-200',
    vendor: 'bg-indigo-100 text-indigo-800 dark:bg-indigo-900 dark:text-indigo-200',
    other: 'bg-gray-100 text-gray-800 dark:bg-gray-800 dark:text-gray-200'
}

// Helper function to get category color
export function getCategoryColor(category: string | undefined): string {
    if (!category) return CATEGORY_COLORS.other
    return CATEGORY_COLORS[category] || CATEGORY_COLORS.other
}

// Helper function to get category label
export function getCategoryLabel(category: string | undefined): string {
    if (!category) return 'Other'
    const found = TEMPLATE_CATEGORIES.find(c => c.value === category)
    return found?.label || category
}

File: backend/app/schemas/template.py
"""
Template Pydantic schemas for request/response validation.
"""
from pydantic import BaseModel, Field, UUID4
from typing import Optional, Dict, Any, List
from datetime import datetime


# Placeholder definition schema
class TemplatePlaceholder(BaseModel):
    """Schema for template placeholder definition."""
    key: str = Field(..., description="Placeholder key, e.g., {{COMPANY_NAME}}")
    label: str = Field(..., description="Human-readable label")
    type: str = Field("text", description="Input type: text, number, date, etc.")
    required: bool = Field(True, description="Whether this field is required")
    default: Optional[str] = Field(None, description="Default value")


# Base schema
class TemplateBase(BaseModel):
    """Base template schema."""
    name: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    content: str = Field(..., min_length=1)
    category: Optional[str] = None


# Schema for creating a template
class TemplateCreate(TemplateBase):
    """Schema for creating a new template."""
    placeholders: List[TemplatePlaceholder] = []
    metadata: Dict[str, Any] = {}


# Schema for updating a template
class TemplateUpdate(BaseModel):
    """Schema for updating template information."""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    content: Optional[str] = Field(None, min_length=1)
    category: Optional[str] = None
    placeholders: Optional[List[TemplatePlaceholder]] = None
    is_active: Optional[bool] = None
    metadata: Optional[Dict[str, Any]] = None


# Schema for template in database (response)
class Template(TemplateBase):
    """Schema for template response."""
    id: UUID4
    placeholders: List[Dict[str, Any]] = []
    is_active: bool
    metadata: Dict[str, Any] = {}
    created_by: UUID4
    created_at: datetime
    updated_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


# Schema for template list item
class TemplateListItem(BaseModel):
    """Schema for template list item."""
    id: UUID4
    name: str
    category: Optional[str] = None
    is_active: bool
    created_at: datetime
    
    class Config:
        from_attributes = True


# Schema for generating contract from template
class GenerateFromTemplate(BaseModel):
    """Schema for generating a contract from template."""
    template_id: UUID4
    title: str = Field(..., min_length=1, max_length=500)
    values: Dict[str, str] = Field(..., description="Placeholder values")
    description: Optional[str] = None
    metadata: Dict[str, Any] = {}
